### PR TITLE
WIP Connection queries for Relayer

### PR DESCRIPTION
**Pull request by adizere**
Closes: #130 

## Description

______

For contributor use:

- [ ] Unit tests written
- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: modules/src/ics03_connection/query.rs [hunk 1]
```diff
@@ -105,6 +119,24 @@ impl IdentifiedConnectionEnd {
     }
 }
 
+fn proto_unmarshal(bytes: Vec<u8>) -> Result<ConnectionEnd, error::Error> {
+    let buf = Bytes::from(bytes);
+    let decoded = proto_ibc_connection::ConnectionEnd::decode(buf);
+    match decoded {
+        Ok(conn) => {
+            let client_id = ClientId::from_str(&conn.client_id).unwrap();
```
Comment (ancazamfir) [inline modules/src/ics03_connection/query.rs:127]: We should `.map_err` here (also for the other `unwrap()`s) and send the error up. For a command like this:
`cargo run --bin relayer -- -c ./relayer/relay/tests/config/fixtures/relayer_conf_example.toml query connection end chain_A connectionnotthere` we should get a friendlier output.

File: modules/src/ics03_connection/query.rs [hunk 2]
```diff
@@ -105,6 +119,24 @@ impl IdentifiedConnectionEnd {
...
+            let connection_end = ConnectionEnd::new(client_id, counterparty, conn.versions).unwrap();
+            Ok(connection_end)
+        },
+        Err(e) => {
+            println!("Error proto un-marshall: {:?}", e);
+            todo!()
+        }
+    }
+}
+
```
Comment (ancazamfir) [inline modules/src/ics03_connection/query.rs:139]: You can remove the `amino` func below.

**Comment by adizere** [comment]
### Before merging:


Because of the code which `prost` generates (based on the .proto file), we had to disable `unused_qualifications` warnings [here](https://github.com/informalsystems/ibc-rs/blob/949ad61e38e1bee2638e858e08bc011ce8d64e4b/modules/src/lib.rs). 
The explicit warning originates from the .rs file generated by prost:

```
  Compiling relayer-modules v0.0.1 (/Users/adi/Hammers/ibc-rs/modules)
error: unnecessary qualification
  --> /Users/adi/Hammers/ibc-rs/target/debug/build/relayer-modules-4a039a021af40008/out/connection.rs:12:13
   |
12 |     pub id: std::string::String,
   |             ^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> modules/src/lib.rs:9:5
   |
9  |     unused_qualifications,
   |     ^^^^^^^^^^^^^^^^^^^^^
```


We have two options:

1. keep the warnings disabled, or
2. re-enable the warnings and bypass the prost compilation step, i.e., we put in our codebase not the .proto files but the final .rs files, modified by hand to pass the `unused_qualifications` warning.

File: modules/src/ics03_connection/exported.rs [hunk 3]
```diff
@@ -42,6 +42,15 @@ impl State {
             Self::Open => "OPEN",
         }
     }
+
+    pub fn from_i32(nr: i32) -> Self {
```
Comment (adizere) [inline modules/src/ics03_connection/exported.rs:46]: @ancazamfir was not sure this is accurate, please double-check.

File: modules/src/ics03_connection/connection.rs [hunk 4]
```diff
@@ -25,6 +29,34 @@ impl ConnectionEnd {
...
+    pub fn set_state(&mut self, new_state: State) {
+        self.state = new_state;
+    }
+
+    pub fn from_proto_connection(pc: proto_connection::ConnectionEnd) -> Result<Self, Error> {
+        // The Counterparty field is an Option, may be missing.
+        match pc.counterparty {
+            Some(cp) => {
+                let mut conn = ConnectionEnd::new(
+                    ClientId::from_str(&pc.client_id).unwrap(),
```
Comment (ancazamfir) [inline modules/src/ics03_connection/connection.rs:42]: Could we get rid of the `unwrap()`s? If the user issues a query command and gets back a few pages of traces it does not help. I had this comment earlier but I guess it got lost. Try to query with non-exsiting connection id.

File: modules/src/ics23_commitment/mod.rs [hunk 5]
```diff
@@ -32,6 +32,10 @@ impl CommitmentProof {
 */
 
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-pub struct CommitmentPrefix;
+pub struct CommitmentPrefix(Vec<u8>);
```
Comment (ancazamfir) [inline modules/src/ics23_commitment/mod.rs:35]: Wondering if we should have a formatter for this. The output of the query shows: `prefix: CommitmentPrefix([112, 114, 101, 102, 105, 120]) }` (decimal), in the genesis and curl it's `cHJlZml4` (base64), ascii is `"prefix"`

**Comment by adizere** [comment]
> I hate to be "that guy" but cherry-picking the proto files for each use-case will become a hassle very quickly. I tried importing them all to see the interconnectedness and in my experience, it's worth keeping proto definitions together.
> 
> See the ibc-proto crate for the results or https://github.com/informalsystems/ibc-proto for the source code.
> 
> I suggest we use that crate or at least import its findings for a more complete proto experience.

Module `ibc_proto::ibc` is private [E0603] :( can we have a quick fix for this @greg-szabo ?

**Comment by greg-szabo** [comment]
Why would you need `ibc_proto::ibc`? It's redundant.
`use ibc_proto::connection` or similar works. Check the documentation: https://docs.rs/ibc-proto/0.1.0/ibc_proto/

File: modules/src/ics03_connection/connection.rs (hunk 4 above)
Comment (greg-szabo) [reply modules/src/ics03_connection/connection.rs:49]: I'd like to make this into a From trait, if possible.

Comment (ebuchman) [reply modules/src/ics03_connection/connection.rs:49]: Also I guess this should be verifying all the fields, not just 2? And should we really have an error per field or maybe start with a more general InvalidConnectionEnd error of some kind and include what's missing as a string? Not sure ...

File: modules/src/ics03_connection/exported.rs (hunk 3 above)
Comment (ancazamfir) [reply modules/src/ics03_connection/exported.rs:46]: Looks good.

File: modules/src/ics03_connection/connection.rs (hunk 4 above)
Comment (adizere) [reply modules/src/ics03_connection/connection.rs]: > I'd like to make this into a From trait, if possible.

Either option could work, the question is if we want to keep validation inside the `from` or outside. The current code uses the `FromStr` trait, with the signature `fn from_str(s: &str) -> Result<Self, Self::Err>` so we can do validation inside there and return error if validation fails. With `From`, the sig is `fn from(T) -> Self` so we'd have to do validation after calling `from`. Is it generally speaking, or specifically in this case, better to do validation outside of `from`?

> Also I guess this should be verifying all the fields, not just 2?

Agree this deserves a discussion to find a more principled approach.

File: modules/src/ics23_commitment/mod.rs (hunk 5 above)
Comment (andynog) [reply modules/src/ics23_commitment/mod.rs:36]: Implemented a fmt::Debug trait to properly format the CommitmentPrefix output in this [commit](https://github.com/informalsystems/ibc-rs/pull/136/commits/f8c0721250f770bc5bdd67b8cdd6ea92532369c6)

**Pull request merged**